### PR TITLE
OCPBUGS-18971: limit number of simultaneous client requests

### PR DESCRIPTION
**Pull request by simonpasquier**
Another implementation for OCPBUGS-18971.

### CODE REVIEW

File: pkg/client/limiter.go
Comment (dgrisonnet) [inline pkg/client/limiter.go:39]: note: we will have to change the registry once https://github.com/kubernetes-sigs/prometheus-adapter/pull/599 is merged and rebased

**Review by dgrisonnet — COMMENTED** [review]
This looks fine to me, but I am still pretty concerned about allowing customers to use VPA with prometheus-adapter. It might have side-effect that they don't expect and disrupt their workload.

File: pkg/client/limiter_test.go
Comment (machine424) [inline pkg/client/limiter_test.go:41]: Knowing that `c.Do(ctx2, "GET", "/nonblocking", nil)` would never reach this, why keeping it? to make the test easier to debug?

File: pkg/client/limiter_test.go
Comment (machine424) [inline pkg/client/limiter_test.go:83]: we can also add a check `total.Load() == maxConcurrentRequests` in here that proves that will also prove that the query doesn't reach the server and also that the equality in line 71 wasn't just transient.

**Review by machine424 — CHANGES_REQUESTED** [review]

File: pkg/client/limiter_test.go
Comment (simonpasquier) [reply pkg/client/limiter_test.go:41]: To ensure that the test catches the issue when/if `c.Do(ctx2, "GET", "/nonblocking", nil)` isn't blocked on the client side.

File: pkg/client/limiter_test.go
Comment (simonpasquier) [reply pkg/client/limiter_test.go:76]: since it returns an error, the first returned value is irrelevant IMHO.

File: pkg/client/limiter_test.go
Comment (simonpasquier) [reply pkg/client/limiter_test.go:83]: it should be detected by the Do() request returning without error.

File: pkg/client/limiter.go
Comment (simonpasquier) [reply pkg/client/limiter.go]: On second thought, I've added a max_requests metric so we can even alert on the prometheus adapter being throttled (e.g. inflight/max == 1 for a significant time).

File: pkg/client/limiter_test.go
Comment (machine424) [reply pkg/client/limiter_test.go:41]: without the `/nonblocking` , in that case the test would be stuck as long as unblock isn't closed.
But ok this would `make the test easier to debug`

File: pkg/client/limiter_test.go
Comment (machine424) [reply pkg/client/limiter_test.go:83]: The client/Do function may be faulty, it's better to check from server side as well. (+ check that the `for total.Load() != maxConcurrentRequests` check wasn't transient)

File: pkg/client/limiter_test.go
Comment (machine424) [reply pkg/client/limiter_test.go:76]: Just a safeguard, in case the receiver doesn't start by checking the error.

File: pkg/client/limiter.go
Comment (machine424) [inline pkg/client/limiter.go:29]:
```suggestion
// getting metrics of all pods in all namespaces would translate into (2 x namespace number) queries to the
```

File: pkg/client/limiter.go
Comment (machine424) [inline pkg/client/limiter.go:31]:
```suggestion
// In the worst case, the Prometheus pods listening sockets queues can hit their limits leading
```

File: pkg/client/limiter.go
Comment (machine424) [inline pkg/client/limiter.go:36]: no power of 2, I'm disappointed :)

At worst a Prometheus pod will serve both prom-adapter (if the other is down or LB missed it) or even 3 or 4 in some weird rolling scenarios which results in 4XX connections < --web.max-connections=512, which leaves room in that queue and the other ones for the other clients.

**Review by machine424 — CHANGES_REQUESTED** [review]

**Comment by machine424** [comment]
(just some nits especially if we don't touch the code in the future, we don't have to worry about regressions)

File: pkg/client/limiter_test.go
Comment (simonpasquier) [reply pkg/client/limiter_test.go:76]: not sure to understand :)

**Comment by simonpasquier** [comment]
/hold
need to fix some nits.

File: pkg/client/limiter_test.go
Comment (machine424) [reply pkg/client/limiter_test.go:83]: it spawns maxConcurrentRequests + 1 = 101 (for i := 0; i < maxConcurrentRequests+1; i++) IIUC.

File: pkg/client/limiter.go
Comment (machine424) [reply pkg/client/limiter.go:36]: Yes, had some similar cases as well while testing, didn't dig into how LB is set.
I think 100 is a good value.

File: pkg/client/limiter_test.go
Comment (jan--f) [inline pkg/client/limiter_test.go:85]: Just for my edification, why do the 101st request separately?

Comment (simonpasquier) [reply pkg/client/limiter_test.go:85]: because it's easier to understand what the test does + L82 ensures that the first 100 requests have hit the server.

**Comment by simonpasquier** [comment]
With @machine424, we checked on a live cluster with 1000 user namespaces + pods with and without the PR.

8 concurrent pods.metrics list clients with this PR:

```
thread 7 (last 10 calls): avg: 12.332307567s, max: 17.173590539s
thread 6 (last 10 calls): avg: 7.09612641s, max: 14.421384552s
thread 2 (last 10 calls): avg: 9.685171344s, max: 14.058853074s
thread 0 (last 10 calls): avg: 8.964397842s, max: 13.960817837s
thread 1 (last 10 calls): avg: 7.854924695s, max: 12.632421842s
thread 4 (last 10 calls): avg: 8.987394292s, max: 14.40615962s
thread 3 (last 10 calls): avg: 8.558045975s, max: 11.919687794s
thread 5 (last 10 calls): avg: 7.505023338s, max: 13.791409648s
```

4 concurrent pods.metrics list clients without this PR:

```
thread 2 (last 10 calls): avg: 10.388223139s, max: 18.269492054s
thread 1 (last 10 calls): avg: 15.003234702s, max: 37.633859814s
thread 0 (last 10 calls): avg: 15.600047318s, max: 37.038546625s
thread 2 (last 10 calls): avg: 16.452837928s, max: 31.755846445s
```

The PR provides better request latency from a client standpoint. Without the PR, increasing the number of pods.metrics clients to 8 triggers liveness probe failures.

**Comment by simonpasquier** [comment]
/hold cancel

it should be good for another round of review now.

**Pull request merged**
